service/src/main/java/com/codeforcommunity/processor/ProtectedSiteProcessorImpl.java

```diff
@@ -436,15 +441,36 @@ public void editSite(JWTData userData, int siteId, EditSiteRequest editSiteReque
   }
 
   @Override
-  public void addSites(JWTData userData, AddSitesRequest addSitesRequest) {
+  public void addSites(JWTData userData, String addSitesRequest) {
     assertAdminOrSuperAdmin(userData.getPrivilegeLevel());
 
-    addSitesRequest
-        .getSites()
-        .forEach(
-            newSite -> {
-              addSite(userData, newSite);
-            });
+    List<AddSiteRequest> addSiteRequests = this.parseCSVString(addSitesRequest);
+
+    addSiteRequests.forEach(siteRequest -> addSite(userData, siteRequest));
+  }
+
+  /**
+   * Parses the given CSV string containing data on a site and site entry to a list of
+   * AddSiteRequests.
+   *
+   * @param sitesCSV CSV string to parse
+   * @throws HandledException if the given CSV string cannot be parsed properly
+   * @return the parsed list of AddSiteRequests
+   */
+  private List<AddSiteRequest> parseCSVString(String sitesCSV) throws HandledException {
+    try {
+      CsvMapper mapper = new CsvMapper();
+      CsvSchema schema = CsvSchema.emptySchema().withHeader();
+      MappingIterator<AddSiteRequest> sitesIterator =
+          mapper.readerFor(AddSiteRequest.class).with(schema).readValues(sitesCSV);
+      List<AddSiteRequest> addSiteRequests = sitesIterator.readAll();
+      for (AddSiteRequest addSiteRequest : addSiteRequests) {
+        addSiteRequest.validate();
+      }
+      return addSiteRequests;
+    } catch (HandledException | IOException e) {
```

Comment:
❓ Is there a specific reason why we catch an IOException to throw an InvalidCSVException? I feel like if something happens with I/O here, it may not always be due to a bad CSV file, unless this is how Jackson signals such a case.

Reply:
Also curious about this ^^

Reply:
Yeah there are tons of reasons why an IOException could be thrown in general (note that the documentation for the two methods we have here that could throw an IOException, [readValues(String)](https://fasterxml.github.io/jackson-databind/javadoc/2.7/com/fasterxml/jackson/databind/ObjectReader.html#readValues(java.lang.String)) and [readAll()](https://fasterxml.github.io/jackson-databind/javadoc/2.7/com/fasterxml/jackson/databind/MappingIterator.html#readAll()), don't actually document when it can be thrown so there's no real way to know exactly when it can be thrown without digging through their code). The reason why I did this in this case was because we're reading from a string (rather than something like a file or an input stream) so that there's only so many reasons why an IOExeption could be thrown, most (if not all) of which is probably due to an invalid CSV file.